File: invocation-patterns/javascript/http_endpoint.js
// Copyright © 2026 Sthenos Security. All rights reserved.
// ============================================================================
// INVOCATION PATTERNS TEST — Case 1: External Endpoint (JavaScript)
//
// All functions are REACHABLE via Express HTTP routes.
// Variables are ATTACKER_CONTROLLED (user input flows to sink).
//
// Expected:
//   app_reachability = REACHABLE
//   taint_verdict    = ATTACKER_CONTROLLED
// ============================================================================
'use strict';
const express = require('express');
const { execSync } = require('child_process');
const path = require('path');
const fs = require('fs');
const http = require('http');

const app = express();
app.use(express.json());


// ── CWE-89: SQL Injection via HTTP endpoint ────────────────────────────

app.get('/case1/sqli', (req, res) => {
    // REACHABLE + ATTACKER_CONTROLLED: req.query → SQL
    const name = req.query.name;
    const db = require('better-sqlite3')(':memory:');
    db.exec(`SELECT * FROM users WHERE name = '${name}'`);
    res.json({ status: 'ok' });
});


// ── CWE-78: Command Injection via HTTP endpoint ────────────────────────

app.post('/case1/cmdi', (req, res) => {
    // REACHABLE + ATTACKER_CONTROLLED: req.body → shell
    const cmd = req.body.cmd;
    const output = execSync(cmd).toString();
    res.json({ output });
});


// ── CWE-22: Path Traversal via HTTP endpoint ───────────────────────────

app.get('/case1/path', (req, res) => {
    // REACHABLE + ATTACKER_CONTROLLED: req.query → file read
    const filename = req.query.file;
    const content = fs.readFileSync(path.join('/var/data', filename), 'utf8');
    res.json({ content });
});


// ── CWE-918: SSRF via HTTP endpoint ────────────────────────────────────

app.get('/case1/ssrf', (req, res) => {
    // REACHABLE + ATTACKER_CONTROLLED: req.query → URL fetch
    const url = req.query.url;
    http.get(url, (response) => {
        let data = '';
        response.on('data', chunk => data += chunk);
        response.on('end', () => res.json({ size: data.length }));
    });
});


app.listen(5011, () => console.log('Case 1 JS server on :5011'));

File: invocation-patterns/python/http_endpoint.py
# Copyright © 2026 Sthenos Security. All rights reserved.
# ============================================================================
# INVOCATION PATTERNS TEST — Case 1: External Endpoint
#
# All functions are REACHABLE via HTTP routes.
# Variables are ATTACKER_CONTROLLED (user input flows to sink).
#
# Expected:
#   app_reachability = REACHABLE (call graph finds @app.route → function)
#   taint_verdict    = ATTACKER_CONTROLLED (AI sees request.args → sink)
# ============================================================================
"""Case 1: External HTTP endpoint → vulnerable function."""
from flask import Flask, request, jsonify
import sqlite3
import subprocess
import os

app = Flask(__name__)


# ── CWE-89: SQL Injection via HTTP endpoint ──────────────────────────────

@app.route('/case1/sqli', methods=['GET'])
def case1_sqli():
    """REACHABLE + ATTACKER_CONTROLLED: request param → SQL."""
    user_id = request.args.get('id', '')
    conn = sqlite3.connect(':memory:')
    conn.execute(f"SELECT * FROM users WHERE id = '{user_id}'")
    conn.close()
    return jsonify({'status': 'ok'})


# ── CWE-78: Command Injection via HTTP endpoint ─────────────────────────

@app.route('/case1/cmdi', methods=['POST'])
def case1_cmdi():
    """REACHABLE + ATTACKER_CONTROLLED: request body → shell."""
    cmd = request.json.get('cmd', 'ls')
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    return jsonify({'output': result.stdout})


# ── CWE-22: Path Traversal via HTTP endpoint ────────────────────────────

@app.route('/case1/path', methods=['GET'])
def case1_path_traversal():
    """REACHABLE + ATTACKER_CONTROLLED: request param → file read."""
    filename = request.args.get('file', '')
    path = os.path.join('/var/data', filename)
    with open(path) as f:
        return jsonify({'content': f.read()})


# ── CWE-918: SSRF via HTTP endpoint ─────────────────────────────────────

@app.route('/case1/ssrf', methods=['GET'])
def case1_ssrf():
    """REACHABLE + ATTACKER_CONTROLLED: request param → URL fetch."""
    import urllib.request
    url = request.args.get('url', '')
    data = urllib.request.urlopen(url).read()
    return jsonify({'size': len(data)})


if __name__ == '__main__':
    app.run(port=5010)
